sympy/sets/tests/test_sets.py

```diff
@@ -14,17 +14,30 @@
 
 def test_imageset():
     ints = S.Integers
+    assert imageset(x, x - 1, S.Naturals) is S.Naturals0
+    assert imageset(x, x + 1, S.Naturals0) is S.Naturals
+    assert imageset(x, abs(x), S.Naturals0) is S.Naturals0
+    assert imageset(x, abs(x), S.Naturals) is S.Naturals
+    assert imageset(x, abs(x), S.Integers) is S.Naturals0
+    # issue 16878a
+    r = symbols('r', real=True)
+    assert (1, r) in imageset(x, (x, x), S.Reals)
+    assert 1 + I in imageset(x, x + I, S.Reals)
     raises(TypeError, lambda: imageset(x, ints))
     raises(ValueError, lambda: imageset(x, y, z, ints))
     raises(ValueError, lambda: imageset(Lambda(x, cos(x)), y))
+    raises(ValueError, lambda: imageset(Lambda(x, x), ints, ints))
     assert imageset(cos, ints) == ImageSet(Lambda(x, cos(x)), ints)
     def f(x):
         return cos(x)
     assert imageset(f, ints) == imageset(x, cos(x), ints)
     f = lambda x: cos(x)
     assert imageset(f, ints) == ImageSet(Lambda(x, cos(x)), ints)
     assert imageset(x, 1, ints) == FiniteSet(1)
-    assert imageset(x, y, ints) == FiniteSet(y)
+    assert imageset(x, y, ints) == Intersection({y}, ints)
+    assert imageset((x, y), (1, z), ints*S.Reals) == Intersection(
+        {(1, z)}, ints*S.Reals)
```

Comment:
Should the result here not just be `{(1,z)}`?

Reply:
not unless z is real

Reply:
`ints*S.Reals` is the domain not the codomain e.g
```
In [11]: imageset(x, S.Half, S.Integers)                                                                                          
Out[11]: ∅
```
er... wait that is not what I expected. How should I interpret this?

Reply:
ugh...this should just be {1/2}: plug in an int always get out 1/2. So then `imageset(x, z, Integers)` must also be `{z}`.